trio_run_in_process/run_in_process.py

```diff
@@ -183,12 +186,40 @@ async def open_in_process(
                         raise err
 
                 await proc.wait()
+                logger.debug(
+                    "process %s finished: returncode=%d", proc, proc.returncode
+                )
 
                 nursery.cancel_scope.cancel()
 
-
-@trio_typing.takes_callable_and_args
-async def run_in_process(async_fn: Callable[..., TReturn], *args: Any) -> TReturn:
-    async with open_in_process(async_fn, *args) as proc:
-        await proc.wait()
-    return proc.result
+    async def run(self, async_fn: Callable[..., TReturn], *args: Any) -> TReturn:
+        async with self.open(async_fn, *args) as proc:
+            await proc.wait()
+        return proc.get_result_or_raise()
+
+    async def __aenter__(self) -> "WorkerProcess":
+        self._parent_r, self._child_w = os.pipe()
+        self._child_r, self._parent_w = os.pipe()
+        parent_pid = os.getpid()
+        self._trio_proc = await trio.open_process(
+            get_subprocess_command(self._child_r, self._child_w, parent_pid),
+            pass_fds=(self._child_r, self._child_w),
+        )
+        await self._trio_proc.__aenter__()
+        self._from_child = await trio.lowlevel.FdStream(self._parent_r).__aenter__()
```

Comment:
*if* you run into problems with manually calling `__aenter__/__aexit__` which I think is problematic in some `trio` APIs (but not necessarily these), I think it's fine to do this API as:

```python
@asynccontextmanager
def open_worker_process(...) -> AsyncIterable[WorkerProcess]:
    ...
```